### ISSUE TITLE
Clashing security providers

### ISSUE THREAD

**Issue by originx** [facebook/conceal#149]
On our production we noticed following devices have clashing security providers between Conceal library and another SDK that we use, here are the stacktraces:

`Fatal Exception: java.lang.IllegalStateException
could not encrypt value
.
.
.
Caused by com.facebook.crypto.exception.KeyChainException: Random number generated
       at com.facebook.android.crypto.keychain.SecureRandomFix.tryApplyFixes(SecureRandomFix.java:69)
       at com.facebook.android.crypto.keychain.SharedPrefsBackedKeyChain.getNewIV(SharedPrefsBackedKeyChain.java:80)
       at com.facebook.crypto.CipherHelper.getCipherOutputStream(CipherHelper.java:37)
       at com.facebook.crypto.Crypto.encrypt(Crypto.java:112)
       at de.payback.core.data.persistence.encryption.ConcealEncryption.encrypt(ConcealEncryption.java:41)
.
.
.
Caused by java.lang.SecurityException: new SecureRandom() backed by wrong Provider: class ag.umt.core.a.a$b$b
       at com.facebook.android.crypto.keychain.SecureRandomFix.tryInstallLinuxPRNGSecureRandom(SecureRandomFix.java:189)
       at com.facebook.android.crypto.keychain.SecureRandomFix.tryApplyFixes(SecureRandomFix.java:63)
       at com.facebook.android.crypto.keychain.SharedPrefsBackedKeyChain.getNewIV(SharedPrefsBackedKeyChain.java:80)
       at com.facebook.crypto.CipherHelper.getCipherOutputStream(CipherHelper.java:37)
       at com.facebook.crypto.Crypto.encrypt(Crypto.java:112)
       at de.payback.core.data.persistence.encryption.ConcealEncryption.encrypt(ConcealEncryption.java:41)`

Version of Conceal we use:
1.0.5@aar <- we plan to migrate to newer version after we thoroughly test the migration process to the new 256bit keys

Affected devices info:
OS VERSION: 4.1.2
88% Samsung models:
gt_i8190 56%
gt_i9100 13%
gt_i8190n 6%
Other... 25%
6% Sony models:
lt22i
6% LGE models:
lg_d605

if you need any extra info let me know

**Comment by originx** [comment]
Probably related to:
https://android-review.googlesource.com/#/c/40145/

is it possible to guard against this issue in Conceal at least catch the exception and retry instead of failing/crashing, issue mostly affects  <=4.2 devices and it is bound to happen that in case of more libs that apply the following fix get the same issue:
http://android-developers.blogspot.de/2013/08/some-securerandom-thoughts.html

**Comment by helios175** [comment]
We removed this code some time ago.
There's not random generator registration anymore.
We plug a SecureRandom implementation to the Conceal singleton (or any other instance you want to use), therefore there's no need for hooking into the global registry.

This caused us some trouble with e2e tests because it seems the provider was registered into the JDK classes, loaded by the testing framework itself, and not released after first test. Second test would try to register itself again (as the static flag would be from the newer class). It would register but the check would not find the same class as it was a class loaded from a previous class loader.

We first workarounded it by checking the classname (even if different Class object) but now this is definitely solved.
